src/libaktualizr/uptane/imagerepository.cc

```diff
@@ -219,11 +221,12 @@ void ImageRepository::updateMeta(INvStorage& storage, const IMetadataFetcher& fe
       local_version = -1;
     }
 
+    const auto timestamp_stored_signature{timestamp.signature()};
     verifyTimestamp(image_timestamp);
 
     if (local_version > remote_version) {
       throw Uptane::SecurityException(RepositoryType::IMAGE, "Rollback attempt");
-    } else if (local_version < remote_version) {
+    } else if (local_version < remote_version || timestamp_stored_signature != timestamp.signature()) {
```

Comment:
One more thing I'm thinking of is this:
- the `verifyXXX` functions store the new data into an instance variable, which can be later reused for different tasks.
- but in case when we throw the security exception above - shouldn't we revert that instance value to the previous value ?

I don't know what other impact it may have, it just seems insane to keep a compomised TUF role in the object state for undefined time.

Reply:
I know, but effectively that's how it has been  for ages and covered with many tests :). Changing this logic is a bit tricky and it might impact many subtle use-cases.

Reply:
yes, you're right

Reply:
> I don't know what other impact it may have, it just seems insane to keep a compomised TUF role in the object state for undefined time.

I gotta say, I completely agree. This has been something I've been aware of for at least a couple years now. I think the `verifyXXX` functions should be const.

> Changing this logic is a bit tricky and it might impact many subtle use-cases.

I agree that it it's tricky and may have some unexpected subtleties, but I think it's worth doing right. I suspect we're more likely to uncover bugs than break things. And if we break something that relied on a shaky foundation, better to fix it anyway and do it right.

Reply:
Maybe, a good trade-off would be really to leave this alone and only do the hot-patch which helps to tackle the exact problem we've seen; and later return back to this when the time allows.

Reply:
That makes sense for now.